### PR TITLE
Fix(54284)

### PR DESCRIPTION
**Pull request by navya9singh**
Fixes https://github.com/microsoft/TypeScript/issues/54284

### CODE REVIEW

File: src/services/refactors/moveToFile.ts [hunk 1]
```diff
@@ -915,26 +912,91 @@ function getRangeToMove(context: RefactorContext): RangeToMove | undefined {
...
+                        or
+       function [|add(x: number, y: number): number;
+       function add(x: string, y: string): string;
+       |]
+                        or
+       function [|add(x: number, y: number): number;
+       function add(x: string, y: string): string;
+       |]const a = 1;
+     */
+    if (overloadRangeToMove && (statements[overloadRangeToMove.end].end >= range.end || range.end === statements[overloadRangeToMove.end].end + 1)) {
```
Comment (andrewbranch) [inline src/services/refactors/moveToFile.ts:929]: This `range.end === node.end + 1` stuff doesn’t feel right. Is this intended to capture the possibility of the range including a trailing newline? What if it includes a trailing space and a trailing newline, or two trailing newlines?

Comment (navya9singh) [reply src/services/refactors/moveToFile.ts]: Yes, it was for a trailing newline. I think this change should fix the other possibilities.

File: src/services/refactors/moveToFile.ts [hunk 2]
```diff
@@ -926,15 +926,15 @@ function getRangeToMove(context: RefactorContext): RangeToMove | undefined {
        function add(x: string, y: string): string;
        |]const a = 1;
      */
-    if (overloadRangeToMove && (statements[overloadRangeToMove.end].end >= range.end || range.end === statements[overloadRangeToMove.end].end + 1)) {
+    if (overloadRangeToMove && (statements[overloadRangeToMove.end].end >= range.end || overloadRangeToMove.end === statements.length - 1 || range.end <= statements[overloadRangeToMove.end + 1].getStart())) {
```
Comment (andrewbranch) [inline src/services/refactors/moveToFile.ts:929]: Rewriting these conditions with some substitutions, this should be equivalent:

```ts
range.end <= statements[N].end ||
range.end <= statements[N + 1].getStart()
```

where `N` is `overloadRangeToMove.end`. Since `statements` is in source order, it sure looks like the first condition is fully subsumed by the second. A statement’s start position can’t possibly be smaller than the previous statement’s end position.

File: src/services/refactors/moveToFile.ts [hunk 3]
```diff
@@ -915,26 +912,91 @@ function getRangeToMove(context: RefactorContext): RangeToMove | undefined {
...
+       function add(x: string, y: string): string;|]
+                        or
+       function [|add(x: number, y: number): number;
+       function add(x: string, y: string): string;
+       |]
+                        or
+       function [|add(x: number, y: number): number;
+       function add(x: string, y: string): string;
+       |]const a = 1;
+     */
```
Comment (DanielRosenwasser) [inline src/services/refactors/moveToFile.ts:928]:
```suggestion
    /**
     * function [|add(x: number, y: number): number;
     * function add(x: string, y: string): string;|]
     *                  or
     * function [|add(x: number, y: number): number;
     * function add(x: string, y: string): string;
     * |]
     *                  or
     * function [|add(x: number, y: number): number;
     * function add(x: string, y: string): string;
     * |]const a = 1;
     */
```

File: src/services/refactors/moveToFile.ts (hunk 2 above)
Comment (DanielRosenwasser) [reply src/services/refactors/moveToFile.ts:929]: I do think you still need to check for if `N + 1` is still valid though, right?

Also, aside, can we rename `.end` to `lastStatementIndex` or something more descriptive?

File: src/services/refactors/moveToFile.ts (hunk 3 above)
Comment (DanielRosenwasser) [reply src/services/refactors/moveToFile.ts:928]: and make similar stylistic changes on other comments.

File: src/services/refactors/moveToFile.ts [hunk 4]
```diff
@@ -915,26 +912,91 @@ function getRangeToMove(context: RefactorContext): RangeToMove | undefined {
...
+       function add(x: string, y: string): string;
+       |]
+                        or
+       const a = 1;[|
+       function add(x: number, y: number): number;
+       function add(x: string, y: string): string;|]
+     */
+    if (endingOverloadRangeToMove) {
+        return {
+            toMove: statements.slice(startNodeIndex, endingOverloadRangeToMove.end === statements.length - 1 ? statements.length : endingOverloadRangeToMove.end + 1),
```
Comment (DanielRosenwasser) [inline src/services/refactors/moveToFile.ts:993]: There are a lot of intermediates here. Be aggressive about lifting them into their own variables so that it's more obvious what each one is for and to cut down on the line length.

File: src/services/refactors/moveToFile.ts (hunk 2 above)
Comment (andrewbranch) [reply src/services/refactors/moveToFile.ts:929]: Yes, I was omitting the length check for simplicity. I might write this something like

```ts
const statementAfterLastOverload = overloadRangeToMove && statements[overloadRangeToMove.end + 1];
if (statementAfterLastOverload && range.end <= statementAfterLastOverload()) {
  // ...
}
```

as I think that shows the intent more clearly than the length check.

File: src/services/refactors/moveToFile.ts (hunk 4 above)
Comment (navya9singh) [reply src/services/refactors/moveToFile.ts]: Does this look okay?

**Review by andrewbranch — COMMENTED** [review]
Can you describe the desired behavior in plain English? Is it just “move the start position of the range back to the start of the set of overloads or the statement it intersects, and move the end position of the range forward to the end of the set of overloads or statement it intersects”? If so, I feel like there must be a more concise way to implement this 🤔

**Comment by navya9singh** [comment]
> Can you describe the desired behavior in plain English? Is it just “move the start position of the range back to the start of the set of overloads or the statement it intersects, and move the end position of the range forward to the end of the set of overloads or statement it intersects”? If so, I feel like there must be a more concise way to implement this 🤔

That's exactly the goal. Tried making it more concise now :)

**Review by andrewbranch — APPROVED** [review]
Nice, this is much easier to follow, at least to me. Thanks!
